src/npc_attack.cpp

```diff
@@ -190,22 +190,15 @@ npc_attack_rating npc_attack_spell::evaluate_tripoint(
         const int distance_to_me = rl_dist( source.pos(), potential_target );
         const bool friendly_fire = att == Creature::Attitude::FRIENDLY &&
                                    !source.rules.has_flag( ally_rule::avoid_friendly_fire );
-        // the distance penalty if it were an npc. if there's no npc, it's 0
-        int npc_dist = source.closest_enemy_to_friendly_distance();
-        if( npc_dist == INT_MAX ) {
-            npc_dist = 0;
-        } else {
-            npc_dist = ( npc_dist - 1 ) * 2;
-        }
-        // a distance of 1 is melee distance, so we want distance shifted over a bit
-        const int distance_penalty = std::max( distance_to_me - 1, npc_dist );
-        int attitude_mult = npc_attack_constants::attitude_multiplier.at( att );
-        if( friendly_fire ) {
+        int attitude_mult = 3;
```

Comment:
The whole attitude_mult can be removed. There shouldn't be any situation where you'd want to willingly target a friendly creature vs. simply not doing anything.
This will cause NPCs to cast offensive spells at you anytime there is danger nearby and the NPC can't reach any other targets.

Reply:
there are situations where you might with spells, as they always have aoe as opposed to other attack types which are single target.

Reply:
additionally, npcs will only cast spells at a target if the evaluation is a net positive number.

Reply:
But this is choosing the target for the spell, not checking the creatures around the target. For what you want you'd have to get the creatures around, find the AoE range of the spell/explosive and come up with an heuristic to adjust the multiplier based on the amount of friendlies inside the AoE and their distance to the center. If allied, NPCs should just return 0 if they are set to avoid friendly fire and there is even a chance of hurting an ally.

This should apply to all NPCs, not just allies, otherwise bandits and such would not care about blowing you up while you are completely surrounded, killing all their friends in the process.

Reply:
it's been a while since i've worked on this, but i made a sub function to check individual tripoints that get all added up in a main function that iterates through all of the aoe points in the spell's aoe

Reply:
Oh, nvm then. I was going by the code for melee/ranged attacks. I'll have to check the magic attacks PR and apply this to the grenades/molotovs too.

Reply:
yes, absolutely. one thing i might be wary about is the code that calculates the aoe for spells is much simpler than the one that calculates for grenades, which uses a flood fill. i haven't delved into that part of the code enough to actually know the details, but that's the main part of the grenade stuff that would be important

Reply:
Well, NPCs shouldn't really know exactly how the explosion will behave so using approximate values depending on the explosive should do.

Reply:
ok that's reasonable then